betaflight/betaflight
# Make UBLOX models more configurable

# Nice description of this PR by @ZzyzxTek :

GPS modules running U-BLOX protocol are configured to use a Dynamic Platform Model. Platform models are used to make internal settings that control the way the module processes positional data, in particular the amount of filtering that is applied. U-BLOX provides seven "standard" platform models: PORTABLE, STATIONARY (most filtering), PEDESTRIAN, AUTOMOTIVE, AT_SEA, AIRBORNE_1G, AIRBORNE_2G, AIRBORNE_4G (least filtering)

IMPORTANT: Dynamic Platform Models also specify maximum expected horizontal and vertical velocities. Selecting a model and exceeding its maximums may trigger sanity checks to fail and cause rejected position readings.
See https://content.u-blox.com/sites/default/files/products/documents/u-blox8-M8_ReceiverDescrProtSpec_UBX-13003221.pdf page 37 for model specifications.

Previously:

- Three user selectable models: AIRBORNE, PEDESTRIAN, DYNAMIC, set via `gps_ublox_mode`.
- AIRBORNE normally mapped to AIRBORNE_4G, or could be changed to AIRBORNE_1G by defining `GPS_UBLOX_MODE_AIRBORNE_1G` at compile time.
- For AIRBORNE or PEDESTRIAN, the user selected model was used when acquiring GPS lock and in flight (after lock).
- For DYNAMIC, PEDESTRIAN was used when acquiring GPS lock, then AIRBORNE in flight.
- The default was `gps_ublox_mode = AIRBORNE`.

There are suggestions to use other Dynamic Platform Models. For example, Oscar Liang suggests PORTABLE is best. Discussions can be found from ~2014-2016.
https://oscarliang.com/gps-settings-u-center/
https://portal.u-blox.com/s/question/0D52p0000CMMaTCCQ1/more-documentation-about-dynamic-platform-models
https://groups.google.com/g/drones-discuss/c/r7gREVd3Ilw/m/GXGP0ZM17BoJ
https://github.com/PX4/PX4-Autopilot/issues/1025
https://rtklibexplorer.wordpress.com/2016/07/19/ublox-m8n-vs-m8t/

We would like to establish a current body of experience within the Betaflight community as to which models work best, and under which circumstances.

The intent of this PR is to enable the community to more easily explore the models, as well as the dynamic behavior of using different models for acquisition and flight, but preserve the prior default behavior.

With these changes:

- User selectable models expanded to all seven standard models: PORTABLE, STATIONARY, PEDESTRIAN, AUTOMOTIVE, AT_SEA, AIRBORNE_1G, AIRBORNE_2G, AIRBORNE_4G
- The compile time define GPS_UBLOX_MODE_AIRBORNE_1G is no longer used.
- When acquiring GPS lock, the model set via `gps_ublox_acquire_model` is used.
- In flight (after lock), the model set via `gps_ublox_flight_model` is used.
- The defaults are now `gps_ublox_acquire_model = AIRBORNE_4G` and `gps_ublox_flight_model = AIRBORNE_4G` which maintains the prior default behavior.
- "Dynamic" is always active, the user can set different acquire and flight models, and those models will be used when acquiring GPS lock and in flight, respectively.

When updating:

- If the user had not changed the prior `gps_ublox_mode` from the default, they should not see any change in behavior.
- If the user had specified a `gps_ublox_mode`, they will now get an error if they apply their old diff and make that setting. Set the desired models using the new settings.
- If the user had specified DYNAMIC, use `set gps_ublox_acquire_model = PEDESTRIAN` and `set gps_ublox_flight_model = AIRBORNE_4G` to restore that behavior.
- If the user had compiled their firmware using the `GPS_UBLOX_MODE_AIRBORNE_1G` define, it will no longer have any effect. Set the desired airborne model using the new settings.

Testing suggestions:

- Use GPS passthrough to read the currently configured Dynamic Platform Model -- TBD need to determine the command and response format to use
- After powering the GPS module, check that the expected acquire model is being used.
- After Betaflight indicates the GPS has acquired a fix, check that the expected flight model is being used.
- Change the acquire and flight models using `set gps_ublox_acquire_model = <model>` and `set gps_ublox_flight_model = <model>`. Repeat the above steps to check they are set as expected.
